File: app/src/main/java/me/tsukanov/counter/domain/CounterModule.java
package me.tsukanov.counter.domain;

import androidx.annotation.NonNull;
import dagger.Module;
import dagger.Provides;
import javax.inject.Singleton;
import me.tsukanov.counter.CounterApplication;
import me.tsukanov.counter.R;
import me.tsukanov.counter.infrastructure.BroadcastHelper;
import me.tsukanov.counter.repository.CounterStorage;
import me.tsukanov.counter.repository.SharedPrefsCounterStorage;

@Module
public class CounterModule {

  private final CounterApplication app;

  public CounterModule(CounterApplication app) {
    this.app = app;
  }

  @Provides
  @Singleton
  CounterApplication provideApp() {
    return app;
  }

  @Provides
  @Singleton
  CounterStorage provideCounterStorage(final @NonNull CounterApplication app) {
    return new SharedPrefsCounterStorage(
        app,
        new BroadcastHelper(app),
        (String) app.getResources().getText(R.string.default_counter_name));
  }
}

File: app/src/main/java/me/tsukanov/counter/domain/CounterModule.kt
package me.tsukanov.counter.domain

import dagger.Module
import dagger.Provides
import me.tsukanov.counter.CounterApplication
import me.tsukanov.counter.R
import me.tsukanov.counter.infrastructure.BroadcastHelper
import me.tsukanov.counter.repository.CounterStorage
import me.tsukanov.counter.repository.SharedPrefsCounterStorage
import javax.inject.Singleton

@Module
class CounterModule(private val app: CounterApplication) {

    @Provides
    @Singleton
    fun provideApp(): CounterApplication {
        return app
    }

    @Provides
    @Singleton
    fun provideCounterStorage(app: CounterApplication): CounterStorage<IntegerCounter> {
        return SharedPrefsCounterStorage(
            app,
            BroadcastHelper(app),
            app.resources.getText(R.string.default_counter_name) as String
        )
    }

}
